### PR TITLE
New message client api

### PR DESCRIPTION
**Pull request by OmarMalik**

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -18,15 +18,111 @@ var uuid = require('node-uuid').v4;
...
+        var options = {
+          job: job,
+          endpoint: endpoint,
+          privateKeyPem: privateKeyPem,
+          publicKeyId: publicKeyId,
+          strictSSL: strictSSL
+        };
+        pollServer(options, function(err, results) {
+          if(err) {
+            callback(err);
```
Comment (dlongley) [inline lib/client.js]: Missing return.

File: lib/client.js [hunk 2]
```diff
@@ -18,15 +18,111 @@ var uuid = require('node-uuid').v4;
...
+          }
+          callback(null);
+        });
+      };
 
-var logger = bedrock.loggers.get('app');
+      scheduler.define(jobId, fn);
+
+      api.start(endpoint, interval, function(err) {
```
Comment (dlongley) [inline lib/client.js]: Can we just do: `api.start(endpoint, interval, callback);` ?

File: lib/client.js [hunk 3]
```diff
@@ -18,15 +18,111 @@ var uuid = require('node-uuid').v4;
...
+      });
+    });
+});
+
+/*
+ * Starts a client that polls the endpoint with the specified interval setting.
+ * If a client is already running on the specified endpoint, it will be
+ * restarted with the new time interval.
+ *
+ * @param endpoint The endpoint for the client to poll
```
Comment (dlongley) [inline lib/client.js]: Add period at the end; lowercase `The` => `the`.

File: lib/client.js [hunk 4]
```diff
@@ -18,15 +18,111 @@ var uuid = require('node-uuid').v4;
...
+ * @param interval 1, 5, or 10 minute intervals to poll the endpoint
```
Comment (dlongley) [inline lib/client.js]: Change to `@param interval the number of minutes to wait between polling the endpoint (e.g. 1, 5, or 10).`

File: lib/client.js [hunk 5]
```diff
@@ -18,15 +18,111 @@ var uuid = require('node-uuid').v4;
...
+
+/*
+ * @param endpoint The endpoint used to create the jobId
+ */
+function createJobId(endpoint) {
+  return 'brMessagesClient.getMessages.' + endpoint;
+}
+
+// exposed for testing
+api._createJobId = function(endpoint) {
```
Comment (dlongley) [inline lib/client.js]: Change to `api._createJobId = createJobId;`

File: tests/01-api.js [hunk 6]
```diff
@@ -120,4 +121,55 @@ describe('bedrock-messages-client API requests', function() {
       });
     });
   });
+
+  describe('start API', function() {
+    before(function() {
+      helpers.removeCollections(function() {
+      });
+    });
```
Comment (dlongley) [inline tests/01-api.js]: This looks like it's supposed to be async, but isn't because the callback isn't used properly. Should this be:

```
before(function(done) {
  helpers.removeCollections(done);
});
```
?

The same applies to `after`. Look up the docs on async `before` and `after` for mocha and make sure we're doing this part properly.

**Comment by dlongley** [comment]
Just a few fixes here and then LGTM. I'll merge once the fixes are in.

File: tests/01-api.js (hunk 6 above)
Comment (OmarMalik) [inline tests/01-api.js]: `before(helpers.removeCollections);` seems to work as intended.

Comment (mattcollier) [inline tests/01-api.js]: removeCollections is definitely an async function.  So it should be...

before(function(done) {
  // done should be passed as the callback to the final operation.
  asyncThing(done);
}

On Mon, Apr 18, 2016 at 1:52 PM, Omar Malik <notifications@github.com>
wrote:

> In tests/01-api.js
> <https://github.com/digitalbazaar/bedrock-messages-client/pull/1#discussion_r60103300>
> :
>
> > @@ -120,4 +121,55 @@ describe('bedrock-messages-client API requests', function() {
> >        });
> >      });
> >    });
> > +
> > +  describe('start API', function() {
> > +    before(function() {
> > +      helpers.removeCollections(function() {
> > +      });
> > +    });
>
> before(helpers.removeCollections); seems to work as intended.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/digitalbazaar/bedrock-messages-client/pull/1/files/2da224b1f14a87aa592522ffc89a1dd75c569e7e#r60103300>
>



-- 
*Matt Collier*

**Comment by dlongley** [comment]
Add configurable endpoint for changing `messages-client/settings` here.

File: lib/client.js [hunk 7]
```diff
@@ -15,18 +15,152 @@ var brMessages = require('bedrock-messages');
...
+      if(!!endpoints) {
+        clients = clients.filter(function(client) {
+          return endpoints.indexOf(client.endpoint) >= 0;
+        });
+      }
+
+      async.each(clients, function(client, callback) {
+        api.start(client.endpoint, interval, callback);
+      },
+      function(err) {
```
Comment (davidlehn) [inline lib/client.js]: Combine with above: "}, function(err) {".

File: lib/client.js [hunk 8]
```diff
@@ -15,18 +15,152 @@ var brMessages = require('bedrock-messages');
...
+  /**
+   * Restarts message clients on the specified endpoints with
+   * the specified interval. If no endpoints specified, all clients will
+   * be restarted.
+   *
+   * @param interval the number of minutes to wait between polling
+   * the endpoint (e.g. 1, 5, or 10)
+   * @param endpoints array of endpoints for clients to be restarted
+   */
+  app.post(config['messages-client']['settings-endpoint'],
```
Comment (davidlehn) [inline lib/client.js]: Should have a body content validator.

File: lib/client.js [hunk 9]
```diff
@@ -15,18 +15,152 @@ var brMessages = require('bedrock-messages');
...
 var scheduler = require('bedrock-jobs');
 var uuid = require('node-uuid').v4;
+require('bedrock-express');
 
 require('./config');
 
+var api = {};
+module.exports = api;
+
+var logger = bedrock.loggers.get('app');
```
Comment (davidlehn) [inline lib/client.js]: Doesn't look like this is used at the moment.

File: lib/client.js [hunk 10]
```diff
@@ -15,18 +15,152 @@ var brMessages = require('bedrock-messages');
...
+          endpoint: endpoint,
+          privateKeyPem: privateKeyPem,
+          publicKeyId: publicKeyId,
+          strictSSL: strictSSL
+        };
+        pollServer(options, function(err, results) {
+          if(err) {
+            return callback(err);
+          }
+          callback(null);
```
Comment (davidlehn) [inline lib/client.js]: Don't need to specify null here.

**Comment by davidlehn** [comment]
Added some more comments.  The content validator is a bit harder to add but would help avoiding bad input.  When is auth going to be added for this?  It's just an open control point right now.

**Comment by OmarMalik** [comment]
I'm planning on adding authentication and validation when we are finishing up M9B, around clean-up time. Right now I want to focus on designing and developing the features for this milestone. It shouldn't be too hard to at the end.

**Comment by OmarMalik** [comment]
Moved the settings endpoint to bedrock-messages-client-http.

File: lib/client.js [hunk 11]
```diff
@@ -15,18 +15,111 @@ var brMessages = require('bedrock-messages');
 var request = require('request');
 var scheduler = require('bedrock-jobs');
 var uuid = require('node-uuid').v4;
+require('bedrock-express');
```
Comment (dlongley) [inline lib/client.js]: Why is this here?

File: lib/client.js [hunk 12]
```diff
@@ -15,18 +15,111 @@ var brMessages = require('bedrock-messages');
...
 
 require('./config');
 
+var api = {};
+module.exports = api;
+
+// TODO: Add logging info
```
Comment (dlongley) [inline lib/client.js]: What's this TODO about?

**Comment by dlongley** [comment]
Just a couple of quick comments ... my guess is `bedrock-express` just didn't get removed when the endpoint moved -- and we should remove that TODO or make the text more specific.

File: lib/client.js (hunk 11 above)
Comment (OmarMalik) [inline lib/client.js]: Forgot to remove this.

File: lib/client.js (hunk 12 above)
Comment (OmarMalik) [inline lib/client.js]: I think I was logging some debugging info but then I got rid of it, the TODO is there to remind me to log some relevant info later to help with debugging.

Comment (OmarMalik) [inline lib/client.js]: I think I'll just remove this TODO for now.
